### PR TITLE
Rework i18n mechanism and add IT layout

### PR DESCRIPTION
**Pull request by SukkoPera**
This is the new way of handling keyboard layouts I proposed in #230.

Only this one or that one should be merged, eventually.

~If this is chosen, the DE keymap must be updated before the merge~ EDIT: Updated

### CODE REVIEW

**Comment by NicoHood** [comment]
Its been long time, that i've been working on this, so please help me once more...

Are you missing some files here? This does not look like a huge change, just an addition of italien layout.

**Comment by SukkoPera** [comment]
Please read #230 carefully, it's all there.

**Comment by SukkoPera** [comment]
This PR is finally complete, please consider for merging. I think this way of handling international keyboards is easier to handle.

File: src/KeyboardLayouts/ImprovedKeylayoutsIT.h [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+	KEY_U,          		// u
+	KEY_V,          		// v
+	KEY_W,          		// w
+	KEY_X,          		// x
+	KEY_Y,          		// y
+	KEY_Z,          		// z
+	KEY_IT_EGRAVE|MOD_LEFT_SHIFT|MOD_RIGHT_ALT,  // {
+	KEY_IT_BACKSLASH|MOD_LEFT_SHIFT,    // |
+	KEY_IT_PLUS|MOD_LEFT_SHIFT|MOD_RIGHT_ALT,  // }
+	KEY_IT_IGRAVE|MOD_RIGHT_ALT,	    // ~ (Linux only: This works in X11, in console it's KEY_0|MOD_RIGHT_ALT)
```
Comment (NicoHood) [inline src/KeyboardLayouts/ImprovedKeylayoutsIT.h:171]: Italiens do not have a tilde? O.o

**Comment by NicoHood** [comment]
Hey @SukkoPera thanks for taking the time to rework your PR! I stared at it for quite some time (it's been a long time since I've been working on this library), but your proposal makes absolutely sense to me. The unified keycode table to press a specific key language independently.

Did you test, that all usual characters are still working fine? What about writing a simple test sketch that we can add to the examples? It could type all characters from space to tilde (poor italiens, no tilde :-P). Then it would be easier to add new layouts and also test them.

Could you please try to verify that all (modified) source files contain spaces, not tabs? I imagine they might be mixed (by me years ago), but they are displayed with mixed indention in Github. Doing a bit of beautifying would be nice, if you like to.

Other than that, I'd say: great work! Did you have any chance to invite someone else to try this out? Maybe someone who is also using this library? Would be nice to hear some additional feedback :-)

File: src/KeyboardLayouts/ImprovedKeylayoutsIT.h (hunk 1 above)
Comment (SukkoPera) [reply src/KeyboardLayouts/ImprovedKeylayoutsIT.h:171]: Unfortunately we don't, and we don't have a backtick either. At least Linux has some easy combinations for those (even though it's always annoyed me that console and X11 use different ones, grrrrr!), but on Windows you get into crappy ALT+numpad stuff, which is a nightmare on notebooks :/.

Comment (NicoHood) [reply src/KeyboardLayouts/ImprovedKeylayoutsIT.h:171]: Thatswhy I am using US layout for coding instead of german layout ;-)

**Comment by NicoHood** [comment]
Did you read me last comment? (Just in case you overlooked that)

**Comment by SukkoPera** [comment]
I did and I agree, but I need some time before I can do that, sorry. Will be back to it next week!

**Comment by NicoHood** [comment]
Alright, no problem :-)

**Comment by NicoHood** [comment]
Ping

**Comment by SukkoPera** [comment]
Pong! Sorry but I've been halted by some dental problems. I will work on this in the weekend, I promise :).

**Comment by SukkoPera** [comment]
Example added. It's not the prettiest but seems to be working very well. Feel free to modify it as you prefer.

File: examples/KeymapTest/KeymapTest.ino [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    delay (5);
+    BootKeyboard.press(KEY_ENTER);
+    delay (5);
+    BootKeyboard.release(KEY_ENTER);
+    delay (5);
+    while (!Serial.available())
+      ;
+    char r = Serial.read();
+    while (Serial.available())
+      Serial.read();
```
Comment (NicoHood) [inline examples/KeymapTest/KeymapTest.ino:50]: Please always use brackets. You cannot tell how error prone this is.

File: examples/KeymapTest/KeymapTest.ino [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    Serial.print(' ');
+    BootKeyboard.press(c);
+    delay (5);
+    BootKeyboard.release(c);
+    delay (5);
+    BootKeyboard.press(KEY_ENTER);
+    delay (5);
+    BootKeyboard.release(KEY_ENTER);
+    delay (5);
+    while (!Serial.available())
```
Comment (NicoHood) [inline examples/KeymapTest/KeymapTest.ino:46]: Please use a single row for the semicolon.

File: examples/KeymapTest/KeymapTest.ino [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+  BootKeyboard.begin();
+  BootKeyboard.releaseAll();
+
+  Serial.println ("--- HID Project Keyboard Layout Test ---");
+  Serial.println ("You will see two columns of characters. If the chosen keymap is correct, the two characters");
+  Serial.println ("will be the same on every line. If they are not, try again with a different keymap.");
+  Serial.println ("DO NOT CHANGE WINDOW UNTIL THE SKETCH IS DONE!!!");
+  Serial.println ("Please click in the input box above, make sure it is empty and press ENTER");
+  while (!Serial.available())
```
Comment (NicoHood) [inline examples/KeymapTest/KeymapTest.ino:29]: Please use a single row for the semicolon.

File: examples/KeymapTest/KeymapTest.ino [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  Copyright (c) 2014-2015 NicoHood
+  See the readme for credit to other people.
+
+  Consumer example
+  Press a button to play/pause music player
+
+  You may also use SingleConsumer to use a single report.
+
+  See HID Project documentation for more Consumer keys.
+  https://github.com/NicoHood/HID/wiki/Consumer-API
```
Comment (NicoHood) [inline examples/KeymapTest/KeymapTest.ino:11]: I guess this is outdated. Please remove or add something different.

File: examples/KeymapTest/KeymapTest.ino [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+*/
+
+#include <HID-Project.h>
+
+void setup() {
+  Serial.begin(115200);
+  while (!Serial)
+    ;
```
Comment (NicoHood) [inline examples/KeymapTest/KeymapTest.ino:19]: Please use a single row for the semicolon.

File: examples/KeymapTest/KeymapTest.ino [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    BootKeyboard.release(KEY_ENTER);
+    delay (5);
+    while (!Serial.available())
+      ;
+    char r = Serial.read();
+    while (Serial.available())
+      Serial.read();
+    Serial.print (r);
+    if (r == c) {
+      Serial.print(" OK");
```
Comment (NicoHood) [inline examples/KeymapTest/KeymapTest.ino:53]: That is a nice idea!

**Comment by NicoHood** [comment]
**Very** nice idea! I've added some comments about some cosmetics. I think the PR is ready than. Thanks a lot for looking into that again :-)

**Comment by SukkoPera** [comment]
I've taken care of all your observations and rebased the branch on your master. I think it's ready for merging too :).

**Comment by NicoHood** [comment]
Thanks a lot

**Pull request merged**